[MRG] rework workflows and management section (#6)
bluegenes/2020-gep: content/04.workflows-and-software.md

| Nextflow | https://www.nextflow.io/ | https://github.com/nf-core/sarek | https://www.nextflow.io/docs/latest/getstarted.html |
| Common workflow language | https://www.commonwl.org/ | https://github.com/EBI-Metagenomics/pipeline-v5 | https://www.commonwl.org/user_guide/02-1st-example/index.html |
| Workflow description language | https://openwdl.org/ | https://github.com/gatk-workflows/gatk4-data-processing |  https://support.terra.bio/hc/en-us/articles/360037127992--1-howto-Write-your-first-WDL-script-running-GATK-HaplotypeCaller |

Table: Popular bioinformatics workflow systems, documentation, example workflows, and tutorials.
While we have linked to general tutorials, there may be more relevant tutorials online for your field.
Not all workflow systems are not necessarily exclusive entities: some workflow systems can translate workflows between languages or run tools or tasks written in other languages.
{#tbl:workflows}

The best workflow system to adopt may be the one with a strong and accessible local or online community in your field, somewhat independent of your computational needs.

> The availability of field-specific data analysis code for reuse and modification, as well as community support for new users can  can facilitate the adoption process (CITE CARPENTRIES?).

Reviewer: ```suggestion
The availability of field-specific data analysis code for reuse and modification, as well as community support for new users can facilitate the adoption process (CITE CARPENTRIES?).
```

program, you can ensure all programs will run without issue. Using small, separate environments for your software and building many simple environments to accommodate different steps in your workflow also reduces the amount of time it takes conda to resolve dependency conflicts between different software tools ("solve" an environment). Conda virtual environments can be created and installed either on the command line, or via an environment YAML file, as shown. In this case, the environment file
also specifies which Conda channels to search and download programs from. When specified in a YAML file, conda environments are easily transferable between computers and operating systems. Further, because the version of each package installed in an environment is recorded, workflow reproducibility is enhanced. Although portions of conda may be superceded by alternative solutions [@url:https://github.com/QuantStack/mamba], this model of software installation and management will likely remain.](images/conda_figure.svg){#fig:conda_figure}

> #### Getting started with software management

Reviewer: ah good idea!!

> Using these software management tools within workflow systems is relatively straightforward: first install the manager, then add a system-specific line to each workflow step that specifies information for software installation. With snakemake, for example, each analysis step must contain a link to a software environment (e.g. on of the boxes described in **Figure {@fig:conda_figure}, B**). If the workflow is then executed with the "--use-conda" option, the software specified will be downloaded, installed, used to run that analysis step, and stored for future analysis runs. The specific syntax for using package managers with your chosen workflow system will be available in the documentation.

Reviewer: It's not a requirement for each step to contain a link to a software environment in snakemake. 
You can also specify singularity from within snakemake, so if we're going to mention `--use-conda` specifically, it may be worth mentioning singularity as well or generalizing between the two.

Author: whoops, meant to say this was needed for snakemake + conda. But maybe I'll try to generalize the example.

While package managers and containers greatly increase reproducibility, there are a number of ways to test software before (or without ever) needing to worry about installation.
Some software packages are available as web-based tools and through a series of data upload and parameter specifications, allow the user to interact with a tool that is running on a back-end server.
This approach is ideal for testing a tool prior to installation to determine whether it produces an appropriate or useful output on your data.
Integrated development environments like PyCharm and RStudio can also manage software installation for the user for language-specific tools.

> In our experience, the complete solution for using scientific software involves a combination of these tools and conda for analysis development, and workflow-integration with conda, singularity, or docker for executing full workflows on many samples.

Reviewer: nice!

#### Getting started with software management

> Using these software management tools within workflow systems is relatively straightforward: install the manager if needed, then add a system-specific line to each workflow step that specifies information for software installation. With snakemake, for example, add a link to an environment, system module, or container for the required software to the code for that analysis step (conda environment example in **Figure {@fig:conda_figure}, B**).

Reviewer: ```suggestion
Using these software management tools within workflow systems is relatively straightforward: install the manager if needed, then add a line to each workflow step specifying information for software installation. (see conda environment example in **Figure {@fig:conda_figure}, B**).
```

Reviewer: I think we can cut it down like that...let's give it a try and we can come back to this section if necessary!

> If the workflow is then executed with the "--use-conda", "--use-singularity", or "-use-envmodules" options, the specified environment or container will be downloaded and installed (or system module will be loaded), used to run that analysis step, and stored for future analysis runs. The specific syntax for using package managers with your chosen workflow system will be available in the documentation.

Reviewer: ```suggestion
When the workflow is executed, the specified environment or container will be downloaded and installed (or system module will be loaded), used to run that analysis step, and stored for future analysis runs. The specific syntax for using package managers with your chosen workflow system will be available in the documentation.
```
